### PR TITLE
feat: multi-tenant support

### PR DESCRIPTION
**Pull request by monsieurBoutte**
# Overview

Adds multi-tenant credential support for Workflow Dev Kit. Credentials can now be encrypted client-side and passed safely through workflow serialization boundaries, then decrypted inside steps on the execution host.

  **Problem:** Workflow Dev Kit serializes workflow inputs/outputs for observability. Passing plaintext API keys through `start()` would expose secrets in logs/traces.

  **Solution:** AES-256-GCM encryption wrapper that encrypts credentials before workflow invocation and decrypts them inside steps using a shared secret key (`MUX_AI_WORKFLOW_SECRET_KEY`). 

Implementation also uses Web Crypto APIs to avoid Node.js modules in workflow functions.

---

# How this works on Vercel (serverless-only setup)

There are two phases, both running in Vercel serverless functions:

  1. **Trigger phase** — your API route calls `start()` and passes **encrypted** credentials.
  2. **Execution phase** — the workflow steps run in a separate serverless invocation; this is the **execution host** where decryption happens.

Set `MUX_AI_WORKFLOW_SECRET_KEY` in your Vercel project so every serverless function can decrypt during step execution.

  ```ts
  import { start } from "workflow/api";
  import { encryptForWorkflow, getSummaryAndTags } from "@mux/ai/workflows";

// Trigger phase (API route): encrypt before calling start()
  const workflowKey = process.env.MUX_AI_WORKFLOW_SECRET_KEY!;
  const encryptedCredentials = await encryptForWorkflow(
    {
      muxTokenId: "mux-token-id",
      muxTokenSecret: "mux-token-secret",
      openaiApiKey: "openai-api-key",
    },
    workflowKey,
  );

  // Execution phase happens later in a separate invocation.
  // The workflow runtime decrypts inside steps using MUX_AI_WORKFLOW_SECRET_KEY.
  const run = await start(getSummaryAndTags, [
    "your-asset-id",
    { provider: "openai", credentials: encryptedCredentials },
  ]);
```
  Key point: ciphertext crosses workflow boundaries; plaintext only exists inside the step runtime.

---

  # What was changed

  ### New files
  | File | Purpose |
  |------|---------|
  | `src/lib/workflow-crypto.ts` | `encryptForWorkflow` / `decryptFromWorkflow` helpers (AES-256-GCM) with optional key ID (`kid`) for rotation |
  | `src/lib/workflow-credentials.ts` | Resolution chain: encrypted creds → provider hook → env fallback |
  | `tests/unit/workflow-crypto.test.ts` | Round-trip, wrong-key, malformed payload, and key ID tests |

  ### Modified files

  **Core:**
  - `src/index.ts` — Exports crypto helpers, `setWorkflowCredentialsProvider`, and version from `package.json`
  - `src/types.ts` — `WorkflowCredentials`, `WorkflowCredentialsInput`, `Encrypted<T>` types
  - `src/env.ts` — Mux creds now optional; added `MUX_AI_WORKFLOW_SECRET_KEY`; requires either Mux creds OR workflow key
  - `src/lib/client-factory.ts` — `getMuxCredentialsFromEnv` / `getApiKeyFromEnv` now async with credentials param
  - `src/lib/providers.ts` — Added `resolveLanguageModelConfig` / `resolveEmbeddingModelConfig`; model factories now async
  - `src/lib/url-signing.ts` — `signUrl` now resolves signing context from credentials
  - `src/lib/workflow-crypto.ts` — Uses Web Crypto + `TextEncoder/TextDecoder`; crypto helpers are async
  - `src/lib/image-download.ts` — Uses `Uint8Array` and Web base64 helpers (no `node:buffer`)
  - `src/workflows/moderation.ts` — Hive file source now uses `Uint8Array` buffers
  - `src/lib/workflow-credentials.ts` — Awaits async decrypt

  **Primitives:**
  - `src/primitives/storyboards.ts`, `thumbnails.ts`, `transcripts.ts` — Accept `credentials` for signed URL generation

  **Workflows (all 7):**
  - `summarization.ts`, `chapters.ts`, `burned-in-captions.ts`, `moderation.ts`, `embeddings.ts`, `translate-captions.ts`, `translate-audio.ts`
  - Each threads `options.credentials` to Mux API calls, AI provider instantiation, and URL signing

  **Docs:**
  - `README.md` — Multi-tenant encryption examples + provider hook usage
  - `docs/API.md`, `docs/WORKFLOWS.md` — Updated credential guidance

  # Suggested review order

  1. **`src/lib/workflow-crypto.ts`** — AES-256-GCM implementation, IV/tag handling, `kid` support (WebCrypto)
  2. **`src/lib/workflow-credentials.ts`** — Resolution chain logic (decrypt → provider → env)
  3. **`tests/unit/workflow-crypto.test.ts`** — Crypto edge cases and key rotation tests
  4. **`src/types.ts`** — Type definitions that flow through entire codebase
  5. **`src/lib/client-factory.ts` + `src/lib/providers.ts`** — Async conversion pattern
  6. **One workflow (e.g., `summarization.ts`)** — Credential threading pattern
  7. **`README.md`** — User-facing docs for encryption + provider hook

  # Key design decisions

  | Decision | Rationale |
  |----------|-----------|
  | Encryption at rest, decryption in steps | Secrets never appear in workflow I/O serialization |
  | Web Crypto (no Node modules) | Workflow functions disallow Node.js core modules |
  | Async crypto helpers | SubtleCrypto APIs are async |
  | Fallback chain (encrypted → provider → env) | Flexible for multi-tenant or single-tenant deployments |
  | Versioned payload `{ v: 1, alg, kid?, iv, tag, ciphertext }` | Supports future algorithm upgrades and key rotation |
  | Optional `kid` field | Enables key rotation without breaking existing payloads |
  | Async credential resolution | Providers can fetch from external secrets managers |
  | Backwards compatible | Existing env-var-only usage continues unchanged |

# Testing notes

  ```bash
  # Run crypto unit tests
  npm test -- workflow-crypto

  # Run all unit tests
  npm test
```

Crypto tests cover: round-trip encryption, wrong key rejection, malformed payloads, key ID presence/absence, and backwards compatibility.

### CODE REVIEW

File: src/lib/client-factory.ts [hunk 1]
```diff
@@ -6,54 +6,30 @@ import type {
...
-      elevenlabs: "ELEVENLABS_API_KEY",
-    };
-    throw new Error(
-      `${provider} API key is required. Set ${envVarNames[provider]} environment variable.`,
-    );
-  }
-
-  return apiKey;
+export async function getApiKeyFromEnv(
+  provider: "openai" | "anthropic" | "google" | "hive" | "elevenlabs",
```
Comment (mmcc) [inline src/lib/client-factory.ts]: Should this list get exported from types? Feels like it should also probably be used [here](https://github.com/muxinc/ai/blob/main/src/types.ts#L97), for example? 

Looking further down in the PR, does seem like there's [`SupportedProvider`](https://github.com/muxinc/ai/blob/b622d9f0302734ce95ab40e5825797b6b7a1875a/src/lib/providers.ts#L10)?

File: src/lib/workflow-credentials.ts [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    throw new Error(
+      "Mux credentials are required. Provide encrypted workflow credentials or set MUX_TOKEN_ID and MUX_TOKEN_SECRET environment variables.",
+    );
+  }
+
+  return { muxTokenId, muxTokenSecret };
+}
+
+/** Supported AI/ML provider identifiers for API key resolution */
+type ProviderApiKey = "openai" | "anthropic" | "google" | "hive" | "elevenlabs";
```
Comment (mmcc) [inline src/lib/workflow-credentials.ts]: See above comment on shared type for this sucker.

File: src/lib/workflow-crypto.ts [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+}
+
+function base64ToBytes(value: string, label: string): Uint8Array {
+  if (!value) {
+    throw new Error(`Invalid encrypted payload: missing ${label}.`);
+  }
+  const normalized = value.length % 4 === 0 ? value : value + "=".repeat(4 - (value.length % 4));
+  if (!BASE64_ALPHABET_RE.test(normalized)) {
+    throw new Error(`Invalid encrypted payload: ${label} is not base64.`);
+  }
```
Comment (mmcc) [inline src/lib/workflow-crypto.ts]: Running a regex on what could be a pretty big string here feels a little overly defensive to me given that `atob` is already going to throw an exception if it's not valid base64 (line 80).

File: src/lib/workflow-crypto.ts [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+  if (typeof globalThis.atob !== "function") {
+    throw new TypeError("Base64 decoder is not available in this environment.");
+  }
```
Comment (mmcc) [inline src/lib/workflow-crypto.ts]: What JS runtime could we possibly be running in that doesn't support a basic runtime feature like this?

File: src/lib/workflow-crypto.ts [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+}
+
+/** Converts key to bytes and validates it's exactly 32 bytes (256 bits) */
+function normalizeKey(key: Uint8Array | string): Uint8Array {
+  let keyBytes: Uint8Array;
+  if (typeof key === "string") {
+    try {
+      keyBytes = base64ToBytes(key, "key");
+    } catch {
+      throw new Error("Invalid workflow secret key. Expected 32-byte base64 value.");
```
Comment (mmcc) [inline src/lib/workflow-crypto.ts]: We throw a bunch of descriptive errors above, then catch and eat them and just return a generic error. Explicitly adding the context of the secret key to the error is helpful, but might still be worth exposing what's thrown within `base64ToBytes`

File: src/workflows/translate-audio.ts [hunk 6]
```diff
@@ -374,25 +393,19 @@ export async function translateAudio(
   const s3AccessKeyId = env.S3_ACCESS_KEY_ID;
   const s3SecretAccessKey = env.S3_SECRET_ACCESS_KEY;
 
-  if (!elevenLabsKey) {
-    throw new Error("ElevenLabs API key is required. Provide elevenLabsApiKey in options or set ELEVENLABS_API_KEY environment variable.");
+  if (!isEncryptedPayload(credentials)) {
+    const elevenLabsKey = credentials?.elevenLabsApiKey ?? env.ELEVENLABS_API_KEY;
```
Comment (mmcc) [inline src/workflows/translate-audio.ts]: I'm too far down in this PR now to scroll back up, forgive my laziness here...but doesn't the credentials provider pass through credentials from the ENV too if they're there? Do we really need to check both everywhere as well?

**Review by mmcc — COMMENTED** [review]
Looks really cool, and we should definitely make sure to get a second set of eyes on the stuff that's touching encryption/decryption from a security angle (not because I see something alarming, but because I don't feel confident in my ability to review that well).

File: src/lib/client-factory.ts (hunk 1 above)
Comment (monsieurBoutte) [reply src/lib/client-factory.ts]: yeah good catch

File: src/lib/workflow-crypto.ts (hunk 4 above)
Comment (monsieurBoutte) [reply src/lib/workflow-crypto.ts]: if i recall correctly it was a wrinkle at vercels workflow runtime

File: src/env.ts [hunk 7]
```diff
@@ -21,8 +22,12 @@ function requiredString(description: string, message?: string) {
 const EnvSchema = z.object({
   NODE_ENV: z.string().default("development").describe("Runtime environment."),
 
-  MUX_TOKEN_ID: requiredString("Mux access token ID.", "Required to access Mux APIs"),
-  MUX_TOKEN_SECRET: requiredString("Mux access token secret.", "Required to access Mux APIs"),
+  MUX_TOKEN_ID: optionalString("Mux access token ID.", "Required to access Mux APIs"),
+  MUX_TOKEN_SECRET: optionalString("Mux access token secret.", "Required to access Mux APIs"),
+  MUX_AI_WORKFLOW_SECRET_KEY: optionalString(
```
Comment (daniel-hayes) [inline src/env.ts:27]: Should we include some docs that go through the process of creating a new key, and how to rotate the key? Using `openssl` or Node.js etc. 

Not necessary for this PR, but having a runbook on how we handle key rotation would be great too.

File: docs/API.md [hunk 8]
```diff
@@ -67,7 +67,7 @@ Analyzes video thumbnails for inappropriate content using OpenAI's Moderation AP
   - `maxRetryDelay?: number` - Maximum delay between retries in milliseconds (default: 10000)
   - `exponentialBackoff?: boolean` - Whether to use exponential backoff (default: true)
 
-All credentials (`MUX_TOKEN_ID`, `MUX_TOKEN_SECRET`, `OPENAI_API_KEY`, `HIVE_API_KEY`) are automatically read from environment variables.
+Credentials (`MUX_TOKEN_ID`, `MUX_TOKEN_SECRET`, `OPENAI_API_KEY`, `HIVE_API_KEY`) are read from environment variables by default. For Workflow Dev Kit multi-tenant usage, pass encrypted `credentials` in options (see README).
```
Comment (daniel-hayes) [inline docs/API.md:70]: Is there a way to detect if this code is being run in a workflow automatically? I'm wondering if we can enforce needing to pass encrypted credentials automatically based on the runtime _or_ if a  `MUX_AI_WORKFLOW_SECRET_KEY` var is set. I think documenting this is also fine for now, but if Vercel exposes workflow env variables by default, this could be a nice way to reduce possibility of accidental exposure.

And if we just check for `MUX_AI_WORKFLOW_SECRET_KEY`, we prevent needing to explicitly pass the secret into `encryptForWorkflow`. Open to discussing the tradeoffs though!

File: src/lib/workflow-credentials.ts [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+ *
+ * Resolution order (later sources override earlier):
+ * 1. Credentials from the registered provider
+ * 2. Decrypted credentials (if input is an encrypted payload)
+ *    OR plain credentials object (if input is already decrypted)
+ *
+ * @param credentials - Optional credentials input (encrypted or plain object)
+ * @returns Merged credentials object
+ */
+export async function resolveWorkflowCredentials(
```
Comment (daniel-hayes) [inline src/lib/workflow-credentials.ts:103]: should we reject plaintext payloads if the MUX_AI_WORKFLOW_SECRET_KEY is an env variable is set?

File: src/lib/workflow-crypto.ts (hunk 3 above)
Comment (daniel-hayes) [reply src/lib/workflow-crypto.ts]: good callout - would having a size/length limit suffice?

**Review by daniel-hayes — COMMENTED** [review]
Looking really good!

File: src/env.ts (hunk 7 above)
Comment (monsieurBoutte) [reply src/env.ts:27]: yeah i think that's a good idea

File: docs/API.md (hunk 8 above)
Comment (monsieurBoutte) [reply docs/API.md:70]: that's a great callout. I know there is indeed a way of figuring out whether or not your in a workflow environment. I'll have to confirm with the Vercel team

File: src/lib/workflow-credentials.ts (hunk 9 above)
Comment (monsieurBoutte) [reply src/lib/workflow-credentials.ts:103]: yes! Just added support for that

File: src/lib/workflow-crypto.ts (hunk 3 above)
Comment (monsieurBoutte) [reply src/lib/workflow-crypto.ts]: opt'ed for removing the regex check here

File: src/lib/workflow-crypto.ts (hunk 4 above)
Comment (monsieurBoutte) [reply src/lib/workflow-crypto.ts]: removed 👍

File: src/workflows/translate-audio.ts (hunk 6 above)
Comment (monsieurBoutte) [reply src/workflows/translate-audio.ts]: yup, good spot

File: src/env.ts (hunk 7 above)
Comment (monsieurBoutte) [reply src/env.ts:27]: added 👍

File: docs/API.md (hunk 8 above)
Comment (monsieurBoutte) [reply docs/API.md:70]: added 👍

File: docs/ENCRYPTED-SECRETS.md
Comment (daniel-hayes) [inline docs/ENCRYPTED-SECRETS.md:1]: 🔥

File: src/lib/workflow-credentials.ts [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+ */
+export function setWorkflowCredentialsProvider(provider?: WorkflowCredentialsProvider): void {
+  workflowCredentialsProvider = provider;
+}
+
+/**
+ * Detects whether code is running inside a Workflow Dev Kit runtime.
+ * getWorkflowMetadata throws when invoked outside a workflow.
+ */
+async function isWorkflowRuntime(): Promise<boolean> {
```
Comment (daniel-hayes) [inline src/lib/workflow-credentials.ts:39]: awesome

**Review by daniel-hayes — APPROVED** [review]
Looks good!
